### PR TITLE
layout: Improve the interaction between baseline-offset-of-last-line-in-flow logic and inline absolute hypothetical boxes.

### PR DESCRIPTION
**Pull request by pcwalton**
See commits for details. These changes place the heart icon on Twitter in the right place.

r? @notriddle

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13470)

### CODE REVIEW

File: components/layout/inline.rs [hunk 1]
```diff
@@ -1262,12 +1262,17 @@ impl InlineFlow {
...
+            None => return None,
+            Some(last_line) => last_line,
+        };
+        if (last_line.range.begin().get()..last_line.range.end().get()).all(|index| {
+            self.fragments.fragments[index as usize].is_hypothetical()
+        }) {
+            None
+        } else {
+            Some(last_line.bounds.start.b + last_line.bounds.size.block -
+                 last_line.inline_metrics.depth_below_baseline)
```
Comment (notriddle) [inline components/layout/inline.rs]: In cases like this:

```html
Some text<br><span style="position:absolute">la la la</span>
```

I would expect the last line to be made up entirely of hypothetical boxes (exactly one), so `baseline_offset_of_last_line` should, in this case, return the baseline offset of "Some text", right?

Comment (pcwalton) [inline components/layout/inline.rs]: Oh yes, of course. I'll fix.

**Comment by pcwalton** [comment]
Addressed. r? @notriddle

File: components/layout/inline.rs [hunk 2]
```diff
@@ -1262,13 +1262,16 @@ impl InlineFlow {
...
+        // Find the last line that doesn't consist entirely of hypothetical boxes.
+        for line in self.lines.iter().rev() {
+            if (line.range.begin().get()..line.range.end().get()).any(|index| {
+                !self.fragments.fragments[index as usize].is_hypothetical()
+            }) {
+                return Some(line.bounds.start.b + line.bounds.size.block -
+                            line.inline_metrics.depth_below_baseline)
             }
         }
+        None
```
Comment (notriddle) [inline components/layout/inline.rs]: This is a massive improvement!

**Review by notriddle — APPROVED** [review]

**Comment by notriddle** [comment]
@bors-servo force p=1

See if making homu push a new thing to auto will kick off buildbot again.

**Pull request merged**